public/app/plugins/datasource/prometheus/querycache/QueryCache.ts

```diff
@@ -0,0 +1,238 @@
+import { ArrayVector, DataFrame, DataQueryRequest, dateTime, Field } from '@grafana/data/src';
+import { amendTable, Table, trimTable } from 'app/features/live/data/amendTimeSeries';
+
+import { PromQuery } from '../types';
+
+// dashboardUID + panelId + refId
+// (must be stable across query changes, time range changes / interval changes / panel resizes / template variable changes)
+type TargetIdent = string;
+
+// query + template variables + interval + raw time range
+// used for full target cache busting -> full range re-query
+type TargetSig = string;
+
+type TimestampMs = number;
+
+type StringInterpolator = (expr: string) => string;
+export const defaultPrometheusQueryOverlapWindowSeconds = 10 * 60 * 1000;
+
+interface TargetCache {
+  sig: TargetSig;
+  prevTo: TimestampMs;
+  frames: DataFrame[];
+}
+
+export interface CacheRequestInfo {
+  requests: Array<DataQueryRequest<PromQuery>>;
+  targSigs: Map<TargetIdent, TargetSig>;
+  shouldCache: boolean;
+}
+
+/**
+ * Get field identity
+ * This is the string used to uniquely identify a field within a "target"
+ * @param field
+ */
+export const getFieldIdent = (field: Field) => `${field.type}|${field.name}|${JSON.stringify(field.labels ?? '')}`;
+
+/**
+ * Get target signature
+ * @param targExpr
+ * @param request
+ * @param targ
+ */
+function getTargSig(targExpr: string, request: DataQueryRequest<PromQuery>, targ: PromQuery) {
+  return `${targExpr}|${request.intervalMs}|${JSON.stringify(request.rangeRaw ?? '')}|${targ.exemplar}`;
+}
+
+/**
+ * NOMENCLATURE
+ * Target: The request target (DataQueryRequest), i.e. a specific query reference within a panel
+ * Ident: Identity: the string that is not expected to change
+ * Sig: Signature: the string that is expected to change, upon which we wipe the cache fields
+ */
+export class QueryCache {
+  private overlapWindowInSeconds: number;
+  constructor(overlapInSeconds?: number) {
+    this.overlapWindowInSeconds = overlapInSeconds ?? defaultPrometheusQueryOverlapWindowSeconds;
```

Comment:
`this.queryOverlapWindow`? (also, just match the constructor arg name).

then multiply it here by 1000 to get millis back

Reply:
I just left this in ms